Treasure Chest Game Shuffle

Jaybone25 opened the pull request:
Gives the ability to shuffle treasure chest game keys and chests.   Has all the options of other keysanity modes.

Jaybone25 committed:
Initial Work for CMG Key Shuffle

Start of work.  Numerous issues to still resolve around permanent flags etc.

Jaybone25 committed:
Add Permanent flags for TCG with Keysanity/Keysy

Added permanent flags 0x1A-0x1F to support locks persisting after leaving the game..   Various other cleanups included as well as missing code in Item.py and Hints.py

Jaybone25 committed:
More TCG work in ASM

Manipulating TCG game, to avoid resetting multiple flags,

Jaybone25 committed:
Missed staged file

Committing a missed staged file

jdunn596 committed:
link tcg keys to key count message

Jaybone25 committed:
More ASM Hacks

Remove in room randomization of chest contents in Chest game.   Change GetItemID of the collection delay item for Chest Mini Game NPC to correct ID for Small Key.

Jaybone25 committed:
Add custom text for Treasure Box Man

Changes various text boxes for treasure chest man when he is shuffled

jdunn596 committed:
refactor shuffle code + add keyring (broken atm)

Jaybone25 committed:
Change ASM Hacks to Conditional

Change ASM hacks from testing variants to conditional based off settings.   Added permanent scene flag when item from chestgame has already been purchased to avoid duping item.

Jaybone25 committed:
Recompile to fix conflicts with pull

Recompile.

jdunn596 committed:
update LocationList and key_rings

jdunn596 committed:
fix typo in key ring message

jdunn596 committed:
undo change to LocationList.py

jdunn596 committed:
fix dungeon menu, keyrings, and keys

jdunn596 committed:
update hintlist, itempool, and overworld.json

Jaybone25 committed:
Various Documentation fixes

Added documentation for sceneflags

Jaybone25 committed:
Fix Location List mix up, let lens see shuffled item in tcg chests

Fixed various items locations in locationlist.py.   Allowed for lens to be able to see the item inside the shuffled chests in chest game.

jdunn596 committed:
fix vanilla tcg behavior

jdunn596 committed:
Fix broken shuffle code

jdunn596 committed:
fix dungeon_menu

jdunn596 committed:
fix vanilla chest size/texture

jdunn596 committed:
update chests.c

jdunn596 committed:
update chests.c again

jdunn596 committed:
fix camc compatibility

jdunn596 committed:
fix chest sizes outside of tcg

jdunn596 committed:
refactor chest code for camc

Jaybone25 committed:
Restore Intended vanilla behavior.   Fix text box.

Restore both chests opening as intended when option is set to vanilla behavior.   Edit the shuffled TCG NPC's text box to look a bit nicer.

jdunn596 committed:
optimize chests.c

jdunn596 committed:
set up gui for upcoming options

jdunn596 committed:
add regional, ow, and any dungeon settings

jdunn596 committed:
update dungeon info menu according to tcg shuffle

jdunn596 committed:
update logic fix ice traps in tcg

jdunn596 committed:
update plandomizer.py, update name

jdunn596 committed:
alter dungeon_info and fix ice traps again

jdunn596 committed:
reformat and cleanup diffs

jdunn596 committed:
update unittests and presets

jdunn596 committed:
fix ci errors

fenhl commented on ASM/c/chests.c:
```diff
@@ -39,15 +40,10 @@ void get_chest_override(z64_actor_t *actor) {
-                    // Small chest
-                    size = SMALL_CHEST;
-                }
-                else {
```
This `else` block needs to be restored.

jdunn596 replied:
addressed

fenhl commented on ASM/c/dungeon_info.c:
```diff
@@ -208,11 +211,12 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
         // Draw keys
 
         if (show_keys) {
+
```
Please don't introduce arbitrary whitespace changes.

Jaybone25 replied:
Removed whitespace, however all existing ones had the same white space.   I did not adjust those

fenhl commented on ASM/c/dungeon_info.c:
```diff
@@ -230,8 +234,10 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
                 char count[5] = "O(O)";
-                if (current_keys > 0) count[0] = current_keys + '0';
-                if (total_keys > 0) count[2] = total_keys + '0';
+                if (SHUFFLE_CHEST_GAME) {
```
It would probably be better to only show the chest game row if this setting is on. In any case, this conditional doesn't make sense as it's being applied to all rows.

jdunn596 replied:
That is actually the way I had it before but because of the way the symbol is written to the rom, it will still show when keysy is on. But I see what you mean otherwise and ill fix it

fenhl replied:
You could set the symbol to `2` in keysy and compare it to `1` here.

jdunn596 replied:
oh so in patches.py, Could I refactor the code to say:

```
if world.settings.shuffle_tcgkeys != 'vanilla':
        if 'keysy' in world.settings.shuffle_tcgkeys:
                rom.write_byte(rom.sym('SHUFFLE_CHEST_GAME'), 0x02)
        else:
                rom.write_byte(rom.sym('SHUFFLE_CHEST_GAME'), 0x01)
        # Update Chest Game Salesman to better fit the fact he sells a randomized item
        update_message_by_id(messages, 0x6D, "I seem to have misplaced my\x01keys, but I have a fun item to\x01sell instead.\x04How about \x05\x4110 Rupees\x05\x40?\x01\x01\x1B\x05\x42Buy\x01Don't Buy\x05\x40\x02")
        update_message_by_id(messages, 0x2D, "That's OK!\x01More fun for me.\x02")
        update_message_by_id(messages, 0x6E, "Wait, that room was off limits!\x02")
        update_message_by_id(messages, 0x704C, "I hope you like it!\x02")
```

Sorry for the questions, I'm still learning lol

fenhl replied:
`world.settings.shuffle_tcgkeys == 'keysy'`

jdunn596 replied:
addressed

fenhl commented on ASM/c/dungeon_info.c:
```diff
@@ -478,8 +484,10 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
                 char count[5] = "O(O)";
-                if (current_keys > 0) count[0] = current_keys + '0';
-                if (total_keys > 0) count[2] = total_keys + '0';
+                if (SHUFFLE_CHEST_GAME) {
```
Same thing here.

jdunn596 replied:
Same as my reply above with this

jdunn596 replied:
addressed

fenhl commented on ItemPool.py:
```diff
@@ -612,6 +619,20 @@ def get_pool_core(world):
                 item = get_junk_item()[0] if location.name != 'Hideout 1 Torch Jail Gerudo Key' else 'Small Key Ring (Thieves Hideout)'
 
+        # Treasure Chest Game Key Shuffle
+        elif location.vanilla_item in ['Small Key (Treasure Chest Game)', 'Green Rupee (Treasure Chest Game)', 'Blue Rupee (Treasure Chest Game)', 'Red Rupee (Treasure Chest Game)'] and location.scene == 0x10:
```
```suggestion
        elif location.scene == 0x10 and location.vanilla_item != 'Piece of Heart (Treasure Chest Game)':
```

Jaybone25 replied:
Fixed

fenhl replied:
Did you forget to push the changes?

jdunn596 replied:
addressed

fenhl commented on Patches.py:
```diff
@@ -2112,6 +2120,10 @@ def update_scrub_text(message, text_replacement, default_price, price, item_name
     HEART_CHEST_SMALL = 16
     HEART_CHEST_BIG = 17
+    if world.settings.shuffle_tcgkeys == 'vanilla':
+        item = read_rom_item(rom, 0x71)
```
This could use a comment explaining what it's doing (changing chest game keys to be in brown chests).

jdunn596 replied:
addressed

fenhl commented on Patches.py:
```diff
@@ -2913,3 +2944,5 @@ def patch_ganons_tower_bk_door(rom: Rom, flag):
     var = (0x05 << 6) + (flag & 0x3F)
     bytes = [(var & 0xFF00) >> 8, var & 0xFF]
     rom.write_bytes(0x2EE30FE, bytes)
+
```
More arbitrary whitespace changes.

Jaybone25 replied:
There were actually 3 lines of white space here.   Removed them all.

fenhl commented on SaveContext.py:
```diff
@@ -1045,7 +1053,10 @@ def get_save_context_addresses(self):
         "Rupee"                     : {'rupees' : None},
-        "Rupee (Treasure Chest Game)" : {'rupees' : None},
+        "Green Rupee (Treasure Chest Game)" : {'rupees' : None},
+        "Blue Rupee (Treasure Chest Game)" : {'rupees' : None},
```
These need to be changed from `None` to their actual values, since the values can't be derived from the item names.

Jaybone25 replied:
Done

fenhl commented on SettingsList.py:
```diff
@@ -3005,15 +3005,57 @@ def __init__(self, name, gui_text, min, max, default, step=1,
-            'choice':    'Choose dungeons',
-            'all':       'All dungeons',
-            'random':    'Random dungeons'
+            'choice':    'Choose dungeons/Treasure Chest Game',
```
If you're going to change this, also add Thieves' Hideout here. I'd say it's not necessary in the option names and should probably be mentioned in the tooltip instead.

jdunn596 replied:
addressed

fenhl commented on data/World/Overworld.json:
```diff
@@ -1385,8 +1385,31 @@
         "locations": {
-            "Market Treasure Chest Game Reward": "can_use(Lens_of_Truth)"
+            "Market Treasure Chest Game Salesman": "is_child and at_night",
+            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys == 'vanilla' and can_use(Lens_of_Truth)) or (shuffle_tcgkeys != 'vanilla' and
```
These can be massively simplified.

* You don't need to check for keysy, that's handled by logically giving the keys as starting items.
* You don't need to check for vanilla other than to see if lens is required. Let the logic assume that you can open both chests, which ends up being equivalent.

For example, I'd write the logic for the reward as:

```python
(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth)) and (Small_Key_Treasure_Chest_Game, 6)
```

jdunn596 replied:
addressed

fenhl reviewed (commented):
In `Notes/extra-save-data.md`, also update the note about the total small key count in the unused field to say `Scenes 0x00–0x10`.

fenhl commented on data/World/Overworld.json:
```diff
@@ -1385,8 +1385,31 @@
     {
         "region_name": "Market Treasure Chest Game",
         "scene": "Market Treasure Chest Game",
+        "hint": "MARKET",
```
This changes the treasure chest game to be considered part of the market hint area regardless of where it's shuffled to. I think this is unintuitive due to its inconsistency with other interiors and should be reverted.

Jaybone25 replied:
This was to troubleshoot a seed generation error due to hint region.   We fixed it and forgot to remove it, we'll remove it

jdunn596 replied:
addressed

Jaybone25 committed:
Small Fixes based on Fehnl's Suggestions

Various bug fixes based off comments

fenhl commented on Patches.py:
```diff
@@ -2943,6 +2943,4 @@ def patch_rupee_tower(location, rom: Rom):
-    rom.write_bytes(0x2EE30FE, bytes)
-
-
+    rom.write_bytes(0x2EE30FE, bytes)
```
You deleted one newline too many. By convention, all text files end with exactly one newline.

jdunn596 replied:
addressed

fenhl commented on data/World/Overworld.json:
```diff
@@ -1387,7 +1387,7 @@
         "hint": "MARKET",
         "locations": {
-            "Market Treasure Chest Game Salesman": "is_child and at_night",
+            "Market Treasure Chest Game Salesman": true,
```
```suggestion
            "Market Treasure Chest Game Salesman": "True",
```

jdunn596 committed:
Update data/World/Overworld.json

Co-authored-by: Fenhl <fenhl@fenhl.net>

jdunn596 committed:
various fixes from fenhl's comments

jdunn596 commented:
All issues _should_ be resolved. I'll wait to hear back :D

fenhl commented on ASM/c/dungeon_info.c:
```diff
@@ -110,7 +113,7 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
         int font_height = 11;
         int padding = 1;
-        int rows = 13;
+        int rows = (!SHUFFLE_CHEST_GAME || SHUFFLE_CHEST_GAME == 2) ? 13 : 14;
```
Can be simplified to:

```suggestion
        int rows = SHUFFLE_CHEST_GAME == 1 ? 14 : 13;
```

fenhl commented on ASM/c/dungeon_info.c:
```diff
@@ -201,39 +204,70 @@ void draw_dungeon_info(z64_disp_buf_t *db) {
-            text_print_size(d->name, left, top, font_width);
-        }
-        text_flush_size(db, font_width, font_height, 0, 0);
+        if (!SHUFFLE_CHEST_GAME || SHUFFLE_CHEST_GAME == 2) {
```
You could merge the “then” and “else” blocks since they're so similar, only checking for `SHUFFLE_CHEST_GAME == 1` in the two places where they should differ, e.g. `i < dungeon_count - (SHUFFLE_CHEST_GAME == 1 ? 0 : 1)`. Same thing for more `if` statements below.

fenhl commented on SaveContext.py:
```diff
@@ -1045,7 +1053,10 @@ def get_save_context_addresses(self):
-        "Rupee (Treasure Chest Game)" : {'rupees' : None},
+        "Green Rupee (Treasure Chest Game)" : {'rupees' : 1},
+        "Blue Rupee (Treasure Chest Game)" : {'rupees' : 5},
+        "Red Rupee (Treasure Chest Game)" : {'rupees' : 10},
```
```suggestion
        "Red Rupee (Treasure Chest Game)" : {'rupees' : 20},
```

fenhl commented on data/World/Overworld.json:
```diff
@@ -1386,7 +1386,18 @@
         "locations": {
-            "Market Treasure Chest Game Reward": "can_use(Lens_of_Truth)"
+            "Market Treasure Chest Game Salesman": "True",
+            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth) and (Small_Key_Treasure_Chest_Game, 6))",
```
```suggestion
            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth)) and (Small_Key_Treasure_Chest_Game, 6)",
```

fenhl commented on data/World/Overworld.json:
```diff
@@ -1386,7 +1386,18 @@
-            "Market Treasure Chest Game Reward": "can_use(Lens_of_Truth)"
+            "Market Treasure Chest Game Salesman": "True",
+            "Market Treasure Chest Game Reward": "(shuffle_tcgkeys != 'vanilla' or can_use(Lens_of_Truth) and (Small_Key_Treasure_Chest_Game, 6))",
+            "Market Treasure Chest Game Room 1 Bottom": "(shuffle_tcgkeys != 'vanilla' and (Small_Key_Treasure_Chest_Game, 1))",
```
```suggestion
            "Market Treasure Chest Game Room 1 Bottom": "(Small_Key_Treasure_Chest_Game, 1)",
```

Similar for the other 9 locations.

fenhl commented on ItemPool.py:
```diff
@@ -612,6 +619,20 @@ def get_pool_core(world):
+                    item = get_junk_item()[0] if location.name != 'Market Treasure Chest Game Salesman' else 'Small Key Ring (Treasure Chest Game)'
+                shuffle_item = True
+            elif world.settings.shuffle_tcgkeys == 'remove' and location.vanilla_item == 'Small Key (Treasure Chest Game)':
+                item = get_junk_item()[0]
```
Keysy is implemented by logically starting with all the keys:

```python
world.state.collect(ItemFactory(item))
```

Add that line above this line so `item` isn't changed to a junk item yet.

jdunn596 committed:
Update SaveContext.py

Co-authored-by: Fenhl <fenhl@fenhl.net>

jdunn596 committed:
Update data/World/Overworld.json

Co-authored-by: Fenhl <fenhl@fenhl.net>

jdunn596 committed:
Update data/World/Overworld.json

Co-authored-by: Fenhl <fenhl@fenhl.net>

jdunn596 committed:
revert unneeded change; make changes per fenhl

Jaybone25 commented:
All suggested changes have been made.

jdunn596 committed:
update item_table.c and hints.py

jdunn596 committed:
revert hints.py

Jaybone25 committed:
Fix for chests opening on return trip

Fixes the Chest game chests from opening if you leave them unopened going into the next room.

jdunn596 committed:
simplify shuffle code

jdunn596 committed:
fix ci error

jdunn596 committed:
restrict opening ice traps in tcg

jdunn596 committed:
fix error in last commit

jdunn596 committed:
fix bug in last commit

jdunn596 committed:
restore original ice trap code in get_items.c

Jaybone25 committed:
Fix bug with Room 1 Bottom Chest

Fixed bug where room 1 bottom chest would be open after making 3 room transitions and opening room 2 bottom chest.   Added conditionals to Icetraps to delay ice trap until leaving chest game.

Jaybone25 committed:
Fix Line break Error

Fix error from CI checks

jdunn596 committed:
fix bug with vanilla; adjust ice traps in tcg

fenhl commented:
I'll give this another round of review soonish.

r0bd0g commented:
For some reason, only Room 1 bottom doesn't check for shuffle_tcgkeys != 'vanilla', which seems strange since all the others do.

It's a bit hard to follow so I just want to make sure, how is the logic getting the 6 keys to reach the final reward when chest game is vanilla?

jdunn596 committed:
fix bug in overworld logic for tcg shuffle

jdunn596 commented:
> For some reason, only Room 1 bottom doesn't check for shuffle_tcgkeys != 'vanilla', which seems strange since all the others do. I think it was a suggested change that said to apply it to the other 9 locations, but this wasn't done.
> 
> It's a bit hard to follow so I just want to make sure, how is the logic getting the 6 keys to reach the final reward when chest game is vanilla? If all of the chests are checking for shuffling to not be vanilla, if you're relying on getting the keys from vanilla locations, that won't work. I checked also, maybe since the keys are irrelevant, you decided to logically start with them, but I don't see that anywhere either. So I'm confused how to logic is to reach the final reward in vanilla behaviour.
> 
> I worry a bit if the intended plan is to collect the keys one at a time from their vanilla locations, if the spoiler log really needs to have like 6 spheres dedicated to collecting and using the vanilla keys. I know that vanilla keys in other places do do that, but the vanilla behaviour of these vanilla key locations is a lot different from the usual vanilla keys, so I'm not sure displaying them in the spoiler log one at a time like that is appropriate.

Good catch on the logic file. I have no idea what happened with that but I assume something happened between both Jay and I merging changes but that should now be fixed.

> I really think the issues with ice traps in the treasure chest game ought to be solved.

In regard to ice traps, we decided to just extend fenhl's method to store them in a queue until you leave the scene before they are applied. You can open them from a chest in the minigame. The textbox will display 'You are a FOOL', and you can continue to 
open the rest of the chests available to you. Upon leaving/save warping, all traps in a q will be applied. There's really not much we can do about this until we better understand why the crashes take place. If you have any ideas, I'm open to suggestions.

jdunn596 committed:
add missing single quotes in overworld.json

r0bd0g commented:
I'm not sure whether the intention was to add the settings check back to the one chest that was missing it, or to remove it from the other 9 that still had it. Maybe Fenhl can explain what he intends there?

That doesn't work for the final chest. It lets you through with lens even when it's not vanilla behaviour.

jdunn596 commented:
> I'm not sure whether the intention was to add the settings check back to the one chest that was missing it, or to remove it from the other 9 that still had it. Maybe Fenhl can explain what he intends there?
> 
> That doesn't work for the final chest. It lets you through with lens even when it's not vanilla behaviour. I think you need not vanilla and keys, or vanilla and lens. (Then the settings checks on all the individual chests won't be necessary since who cares if the logic picks up some vanilla keys or loser rupees? But again I want to know what Fenhl intended with the suggestion.)

Yeah that's actually what we had originally but I guess we'll just wait for it to be reviewed again and see what fenhl thinks

jdunn596 committed:
fix another logic bug per r0b

jdunn596 commented:
You were right r0b I tested with a spoiler log and it was indeed accounting for lens for all settings. The adjustment should handle it

mzxrules commented on LocationList.py:
```diff
@@ -1855,12 +1866,12 @@ def shop_address(shop_id, shelf_id):
     ("Gerudo Training Ground MQ Flame Circle Chest",                 ("Chest",        0x0B,  0x0E, None,                            'Small Key (Gerudo Training Ground)',    ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
     ("Gerudo Training Ground MQ Maze Right Central Chest",           ("Chest",        0x0B,  0x05, None,                            'Rupees (5)',                            ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
-    ("Gerudo Training Ground MQ Maze Right Side Chest",              ("Chest",        0x0B,  0x08, None,                            'Rupee (Treasure Chest Game)',           ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
+    ("Gerudo Training Ground MQ Maze Right Side Chest",              ("Chest",        0x0B,  0x08, None,                            'Green Rupee (Treasure Chest Game)',     ("Gerudo Training Ground MQ", "Master Quest", "Chests",))),
```
nitpick, but should this be Rupee (1) (Treasure Chest Game) to match other instances?

fenhl replied:
I think `Rupee (Treasure Chest Game) (1)` would be best as it would work automatically with the starting items code that translates item names in that format into ammo amounts.

mzxrules commented on Patches.py:
```diff
@@ -2384,6 +2400,25 @@ def update_scrub_text(message, text_replacement, default_price, price, item_name
 
+    # Convert temporary flags used for locked doors in Treasure Chest Game to permanent flags namely (0x1A-0x1F)
+    if world.settings.shuffle_tcgkeys != 'vanilla':
+        rom.write_byte(0x33A6070 +15, 0xDF)
```
```suggestion
        rom.write_byte(0x33A607F, 0xDF)
```

repeat for all

jdunn596 committed:
change target in patches.py, rename tcg rupees

jdunn596 commented:
suggested changes have been made in the most recent commit

fenhl commented:
> Maybe Fenhl can explain what he intends there?

They/them pronouns for me please. As for the question itself, I'm going to send a PR once this PR is rebased onto latest Dev, since that's easier than trying to explain it in a review.

jdunn596 committed:
recompile after rebase

fenhl committed:
Various fixes for Treasure Chest Game shuffle

fenhl committed:
Update glitched logic for TCG shuffle

fenhl commented:
Jaybone25#6

jdunn596 committed:
Merge pull request #6 from fenhl/tcg-shuffle-review

Various fixes for Treasure Chest Game shuffle

r0bd0g commented:
Logic-wise the only thing I can really say about this implementation is that technically the chest that isn't assigned as the vanilla location doesn't need all that extra logic.

fenhl commented:
I think the logic is clearer this way, since the chests are arbitrarily assigned rupees and keys as a hack.

r0bd0g commented:
I think you should comment in the logic file which chests actually have the vanilla keys.

r0bd0g commented:
I mean I think the logic files should be aware of the difference between top and bottom somehow. Otherwise it's kind of stealth atm.

cjohnson57 commented:
Lmao
![image](https://user-images.githubusercontent.com/30701749/236648236-d9f9bde1-8e04-47be-8b4f-9b298dd93bc7.png)

Jaybone25 commented:
Unfortunately that line will rarely be seen lol

On Sat, May 6, 2023 at 5:03 PM Caleb Johnson ***@***.***>
wrote:

> Lmao
> [image: image]
> <https://user-images.githubusercontent.com/30701749/236648236-d9f9bde1-8e04-47be-8b4f-9b298dd93bc7.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OoTRandomizer/OoT-Randomizer/pull/1946#issuecomment-1537231136>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWJY3QMJVEEH3K3QZRHRKU3XE3DDNANCNFSM6AAAAAAWMU4VJM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

cjohnson57 reviewed (approved):
One non-blocking issue, if you can fix before I merge that's great but it won't stop me from merging:
If you say "No" to buying his item, you need to reload the scene to be able to talk to him again.

Jaybone25 commented:
> 

Will try, I attempted it before w/o success, but I have an hour to take a look again
